### PR TITLE
plan: support limit pushdown

### PR DESCRIPTION
**Pull request by hanfei19910905**

### CODE REVIEW

File: optimizer/plan/planbuilder_join.go [hunk 1]
```diff
@@ -64,6 +64,7 @@ type joinPath struct {
 	idxDepCount   int // number of paths this table depends on.
 	ordering      *ast.ResultField
 	orderingDesc  bool
+	LimitDesc     *Limit
```
Comment (coocood) [inline optimizer/plan/planbuilder_join.go]: `Desc` is already used as `Descending`, should not be used as Description`.

File: optimizer/resolver.go [hunk 2]
```diff
@@ -746,6 +746,7 @@ func (nr *nameResolver) createResultFields(field *ast.SelectField) (rfs []*ast.R
 		rf.DBName = v.Refer.DBName
 		rf.TableName = v.Refer.TableName
 		rf.Expr = v
+	case *ast.ParenthesesExpr:
```
Comment (coocood) [inline optimizer/resolver.go]: remove this case.

File: optimizer/plan/planbuilder_join.go (hunk 1 above)
Comment (coocood) [inline optimizer/plan/planbuilder_join.go]: This field don't need to be exported as `jionPath` is not exported.

**Comment by shenli** [comment]
Is limit info used in local_rs.go?

**Comment by ngaut** [comment]
@BusyJay PTAL

**Comment by coocood** [comment]
@shenli not yet.

File: optimizer/plan/plan_test.go [hunk 3]
```diff
@@ -264,7 +264,31 @@ func (s *testPlanSuite) TestBestPlan(c *C) {
...
-			best: "Index(t.b)->Fields->Limit",
+			best: "Index(t.b) + Limit(100)->Fields->Limit",
+		},
+		{
+			sql:  "select * from t where a > 0 order by b DESC limit 100",
+			best: "Index(t.b) + Limit(100)->Fields->Limit",
+		},
+		{
+			sql:  "select * from t where a > 0 order by b + a limit 100",
+			best: "Range(t)->Fields->Sort",
```
Comment (coocood) [inline optimizer/plan/plan_test.go]: Change the stringer for `Sort` plan, so we can see limit is pushed down to `Sort`.

**Comment by coocood** [comment]
LGTM

File: optimizer/plan/planbuilder.go [hunk 4]
```diff
@@ -578,6 +585,19 @@ func buildResultField(tableName, name string, tp byte, size int) *ast.ResultFiel
 	}
 }
 
+func pushLimit(p Plan, limit *ast.Limit) {
+	switch x := p.(type) {
+	case *IndexScan:
```
Comment (shenli) [inline optimizer/plan/planbuilder.go]: case *IndexScan, *TableScan:
    limitCount := int64(limit.Offset + limit.Count)
    x.LimitCount = &limitCount

File: optimizer/plan/plans.go [hunk 5]
```diff
@@ -161,6 +163,8 @@ type IndexScan struct {
 	// TableName is used to distinguish the same table selected multiple times in different place,
 	// like 'select * from t where exists(select 1 from t as x where t.c < x.c)'
 	TableName *ast.TableName
+
+	LimitCount *int64
```
Comment (shenli) [inline optimizer/plan/plans.go]: Why use *int64 instead of int64?

File: optimizer/plan/planbuilder.go (hunk 4 above)
Comment (hanfei19910905) [inline optimizer/plan/planbuilder.go]: will fix

File: optimizer/plan/plans.go (hunk 5 above)
Comment (hanfei19910905) [inline optimizer/plan/plans.go]: because pb will ignore it when it is a null pointer.

**Comment by shenli** [comment]
LGTM

**Pull request merged**